### ISSUE TITLE
[Excalibur G770] Possible problem about color zones? 

### ISSUE THREAD

**Issue by kuylar** [Mustafa-eksi/casper-wmi#1]
First of all, thank you for this awesome driver!

After adding my laptop's model number in `casper_dmi_list`, I found out that all zone IDs change the left zone (for example, zone 0x04 changes left and center, 0x05 changes left and right). Also the zone id 0x07 doesn't seem to change anything.

Also disabled the if statement in [L36](https://github.com/Mustafa-eksi/casper-wmi/blob/0c8d2d74853ce56e20433ab81a6f981e7b3068b7/casper-wmi.c#L136) to try with more zone IDs, and ended up with this chart
![image](https://github.com/Mustafa-eksi/casper-wmi/assets/52961639/4650bba9-7fe3-4106-8cfb-b37a2f2207ea)

Does this also happen in other models, or is it special for my hardware? I can provide more info if needed.

I'm on Arch Linux with kernel version 6.7.5-arch1-1, Excalibur G770 (i5-10300H / GTX 1650)

**Comment by Mustafa-eksi** [comment]
> Does this also happen in other models, or is it special for my hardware? I can provide more info if needed.  

It doesn't happen in mine (G900, 10th gen).
Can you send this file: ```/sys/firmware/acpi/tables/DSDT``` either decompiled or not.
This file has the code that the driver communicates to change the keyboard color. The part I need is a device named "GWMI". I suspect there's something different about it but if it doesn't then this bug is probably hardware related.
For more information: https://wiki.archlinux.org/title/DSDT#Recompiling_it_yourself
And thanks for testing the driver.

**Comment by kuylar** [comment]
[Here's (hopefully) the decompiled GWMI device part](https://github.com/Mustafa-eksi/casper-wmi/files/14395829/dsdt.gwmi.dsl.txt)

If that's not the right one, [the .dat and .dsl file](https://github.com/Mustafa-eksi/casper-wmi/files/14395835/dsdt.tar.gz)

**Comment by Mustafa-eksi** [comment]
Our dsdts are nearly the same, and I couldn't find something useful about these things. I can only speculate that this is a firmware malfunction. Was that bug happening on Windows too? If it wasn't then led id may be different, for that case you can try more ids (you can test 32 bit led ids if you update) but this I don't think there is a different id.  
And there is the diff between my DSDT and yours. (< is mine, > is yours)
```
8c8
<  * Disassembly of DSDT, Sun Feb 25 17:35:59 2024
---
>  * Disassembly of dsdt.dat, Sun Feb 25 11:53:46 2024
14c14
<  *     Checksum         0x01
---
>  *     Checksum         0xF9
332c332
<     OperationRegion (GNVS, SystemMemory, 0x9BB63298, 0x07F9)
---
>     OperationRegion (GNVS, SystemMemory, 0x9BB83298, 0x07F9)
1400c1400
<     OperationRegion (OGNS, SystemMemory, 0x9BB59018, 0x00001003)
---
>     OperationRegion (OGNS, SystemMemory, 0x9BB79018, 0x00001003)
3969c3969
<     OperationRegion (MDBG, SystemMemory, 0x9BB42018, 0x00001004)
---
>     OperationRegion (MDBG, SystemMemory, 0x9BB62018, 0x00001004)
4332c4332
<     Name (PNVB, 0x9BB63018)
---
>     Name (PNVB, 0x9BB83018)
```

**Comment by kuylar** [comment]
Just tested zone IDs from 0x00 to 0xFF, and it looks like after 0x0F, the colummn 2 in the image I sent repeats. Before testing with 32 bits though, I've decided to throw the Excalibur Control Center app into ILSpy to check how it handles the LED settings. As I suspected, they set it from 5 -> 4 -> 3 so that's why I never noticed it. Can't find anything about zone 7 not working though, maybe the app never uses it? (`IsKBLEDModule3` being `false`)

![Decompiled Control Center code](https://github.com/Mustafa-eksi/casper-wmi/assets/52961639/4c840a24-4049-41a8-bd15-178ae0f17942)

> notes about the above snippet
> I have no idea where `IsNewKBZone` gets set from, but basically it swaps the zone 3 and 5 inside `SetLEDModeChangeData`. If `IsSetLEDData` is true, saves the currect backlight data to the registry.

Closing as it seems to work as intended
